### PR TITLE
refactor(conf): split code, generate typings

### PR DESCRIPTION
**Pull request by thekip**
# Description

Refactoring of `@lingui/conf` package.  No changes in logic or behavior. This PR makes package clean and maintainable. 

- Code was split into files
- Deprecation migrations extracted from base file and grouped in own files
- Removed `index.d.ts` which is in turn a d.ts snapshot of former index.ts made by someone, some time ago. 
- Build changed to TSC, so typings now produced automatically and they are aligned with code in ./src
- privates is no longer exposed from package. Type definitions not bloated by `defaultConfig`, `exampleConfig` etc
- Introduced type `LinguiConfigNormalized`:
   `LinguiConfig` is used for typing user provided config. Some of the properties might be in various forms. Config package process user provided config, normalize this various forms into one and that shape become `LinguiConfigNormalized` and supposed to use in the library internals
- I wrote new tests, fixed old ones, and fixed all places where Typescript complained about types mismatch

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to Lingui?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

Fixes # (issue)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [ ] I have read the [CONTRIBUTING](https://github.com/lingui/js-lingui/blob/main/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/lingui/js-lingui/blob/main/CODE_OF_CONDUCT.md) docs
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: packages/cli/src/test/compile.test.ts
Comment (thekip) [inline packages/cli/src/test/compile.test.ts:176]: Just found that one of the tests is always "One snapshot written", that's because mockFs wasn't restored before snapshotting.

**Comment by thekip** [comment]
Changed base to main as discussed in [discord](https://discord.com/channels/974702239358783608/1073174343275773992)

File: packages/conf/src/migrations/catalogMigration.ts
Comment (thekip) [inline packages/conf/src/migrations/catalogMigration.ts:69]: I think we can remove it in this PR, because this is going to be released with version 4. And was deprecated in version 3, together with FallbackLanguage migration.

File: packages/conf/src/types.ts
Comment (thekip) [inline packages/conf/src/types.ts:2]: This import is incorrect. conf should not rely on cli package, please see dicussion: 

https://discord.com/channels/974702239358783608/1072883644273795082

**Comment by andrii-bodnar** [comment]
@thekip could you please explain more about code coverage-related changes? The codecov reports a huge coverage decrease

**Comment by thekip** [comment]
> @thekip could you please explain more about code coverage-related changes? The codecov reports a huge coverage decrease

It's not related. The changes just disable coverage gathering by default in the config, and enabling reporting in yarn commands, which should be used in WF. This was done to make local development more convenient by not gathering coverage when you run tests one by one during coding. 

That should not affect coverage anyhow. Probably it's again wrong base chosen by codecov. 
Anyway i reverted this changes they are not related to this PR and was added to commit by mistake. I would create a separate PR with that changes.

File: packages/conf/src/utils/replaceRootDir.ts
Comment (Martin005) [inline packages/conf/src/utils/replaceRootDir.ts:1]: As this function is used only for replacing th "<rootDir>" string in the config file, I think that `config` can be of type `LinguiConfig`

Comment (thekip) [reply packages/conf/src/utils/replaceRootDir.ts:1]: Actually not, it not bind to shape of LinguiConfig. It can take anything, and return anything. And it's not directly compatible with LinguiConfig. This function called in a recursion, and subsequent calls could be with type "string" or array, etc. So if you put `config: LinguiConfig` typescript would complain

Comment (thekip) [reply packages/conf/src/utils/replaceRootDir.ts:1]: try it yourself please.

Comment (thekip) [reply packages/conf/src/utils/replaceRootDir.ts]: done

File: packages/conf/src/migrations/catalogMigration.ts
Comment (thekip) [reply packages/conf/src/migrations/catalogMigration.ts:69]: @Martin005 @andrii-bodnar i'm waiting for an answer here.

**Review by Martin005 — APPROVED** [review]

File: packages/conf/src/migrations/catalogMigration.ts
Comment (thekip) [reply packages/conf/src/migrations/catalogMigration.ts:69]: i will remove it next iteration. Let's merge this. I want to unblock other PRs

**Review by andrii-bodnar — APPROVED** [review]

**Pull request merged**
